### PR TITLE
Issue/6508 google login callbacks

### PR DESCRIPTION
**Pull request by theck13**
### Fix
Add Google API calls and callbacks to the email screen as described in https://github.com/wordpress-mobile/WordPress-Android/issues/6508.

### Test
0. Clear app data.
1. Tap ***Log In*** button.
2. Tap Google login view.
3. Choose Google account.
4. Notice messages in `logcat`.
```
org.wordpress.android.beta D/WordPress-API: Dispatching action: AccountAction-PUSH_SOCIAL_LOGIN
```
```
org.wordpress.android.beta E/WordPress-API: LoginEmailFragment.onSocialChanged: UNKNOWN_USER - Hmm, we can't find a WordPress.com account with this username and password combo. Please double check your information and try again.
```
```
org.wordpress.android.beta E/WordPress-API: LoginEmailFragment.onSocialChanged: USER_EXISTS - This social account is already associated with an existing WordPress.com account.
```

#### Note
In order to hit the `OnAuthenticationChanged` callback , the Google account selected must already be connected to WordPress.  You can connect/disconnect a social account to/from a WordPress account at https://wordpress.com/me/security/social-login/ on the web.  If it is connected, you can set a breakpoint [here](https://github.com/wordpress-mobile/WordPress-Android/pull/6671/files#diff-c0e31e11f2dbf99904de706676fa1602R433).  If it's not connected, you'll receive the `USER_EXISTS ` error message in `OnSocialChanged`.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginEmailFragment.java [hunk 1]
```diff
@@ -75,6 +108,24 @@ protected void setupContent(ViewGroup rootView) {
         autoFillFromBuildConfig("DEBUG_DOTCOM_LOGIN_EMAIL", mEmailInput.getEditText());
         mEmailInput.addTextChangedListener(this);
         mEmailInput.setOnEditorCommitListener(this);
+
+        LinearLayout googleLoginLayout = (LinearLayout) rootView.findViewById(R.id.login_google_layout);
+        googleLoginLayout.setVisibility(BuildConfig.LOGIN_GOOGLE_ENABLED ? View.VISIBLE : View.GONE);
```
[the remark this answers is not in the document]
Comment (theck13) [reply WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginEmailFragment.java]: The `LOGIN_GOOGLE_ENABLED` feature flag was meant to keep the Google login button hidden from release builds while the feature was still in development.  I forgot that's why I created the `feature/google-login` branch.  I'll remove that feature flag and update the base branch of this pull request to `feature/google-login`.

File: WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginEmailFragment.java [hunk 2]
```diff
@@ -241,4 +365,92 @@ public void onAvailabilityChecked(OnAvailabilityChecked event) {
...
+                    }
+                }
+
+                break;
+        }
+    }
+
+    @SuppressWarnings("unused")
+    @Subscribe(threadMode = ThreadMode.MAIN)
+    public void onAuthenticationChanged(OnAuthenticationChanged event) {
```
[the remark this answers is not in the document]
Comment (theck13) [reply WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginEmailFragment.java]: Yes, `OnAuthenticationChanged` is hit.  It is also possible an error will occur during the callback, but the user interface is being handled in subsequent changes.  That's why the `// TODO: Show error screen.` comment was added to this pull request.

**Comment by theck13** [comment]
The comments about [`LoginEmailFragment.java`](https://github.com/wordpress-mobile/WordPress-Android/pull/6671#discussion_r142114082) and [`login_email_screen.xml`](https://github.com/wordpress-mobile/WordPress-Android/pull/6671#discussion_r142122336) actually refer to changes in https://github.com/wordpress-mobile/WordPress-Android/pull/6522.  They were mistakenly listed in this pull request since the base branch was originally `develop` rather than `feature/google-login`.  Order has been restored and only one [comment](https://github.com/wordpress-mobile/WordPress-Android/pull/6671#discussion_r142121056) remains.

Moving the Google login code to a separate fragment is interesting.  I'll think some more about that one and see what that would mean for the subsequent pull requests.

**Comment by theck13** [comment]
I think moving the Google login code to a separate fragment will work.  I propose we address that in a separate pull request after https://github.com/wordpress-mobile/WordPress-Android/pull/6674 is merged to avoid conflicts.  What do you think?

**Comment by hypest** [comment]
> I think moving the Google login code to a separate fragment will work. I propose we address that in a separate pull request after #6674 is merged to avoid conflicts. What do you think?

Works for me! Thanks so much for considering that change!


> The comments about LoginEmailFragment.java and login_email_screen.xml actually refer to changes in #6522. They were mistakenly listed in this pull request since the base branch was originally develop rather than feature/google-login. Order has been restored and only one comment remains.

Ah, I see. Sorry I missed that bit. I didn't follow the work in #6522 that closely and missed it.

Still, I'd suggest we address the [AppPrefs comment](https://github.com/wordpress-mobile/WordPress-Android/pull/6671#discussion_r142114082) regardless, in the new PR. Thanks!

File: WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginEmailFragment.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (hypest) [reply WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginEmailFragment.java]: Hmm, I can't seem to be able to hit it (debugger breakpoints). Tried with 2 different google accounts and I can normally see all the logcat cases you mention in the PR description. Is there some step I'm missing perhaps?

[the remark this answers is not in the document]
Comment (theck13) [reply WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginEmailFragment.java]: To be able to hit `OnAuthenticationChanged`, the Google account must already be connected to WordPress.  If it is connected, you can set a breakpoint [here](https://github.com/wordpress-mobile/WordPress-Android/pull/6671/files#diff-c0e31e11f2dbf99904de706676fa1602R433).  If it's not connected, you'll receive the `USER_EXISTS ` error message in `OnSocialChanged`.

**Comment by theck13** [comment]
I'll move the Google login code to a separate fragment and address the feature flag kerfuffle in a separate technical debt issue after the currently submitted changes are merged.

File: WordPress/src/main/java/org/wordpress/android/ui/accounts/LoginActivity.java [hunk 3]
```diff
@@ -348,8 +348,8 @@ public void loggedInViaUsernamePassword(ArrayList<Integer> oldSitesIds) {
     }
 
     @Override
-    public void helpEmailScreen(String email) {
-        launchHelpshift(null, email, true, Tag.ORIGIN_LOGIN_EMAIL);
+    public void helpEmailScreen(String email, Tag origin) {
```
[the remark this answers is not in the document]
Comment (theck13) [reply WordPress/src/main/java/org/wordpress/android/ui/accounts/LoginActivity.java]: I updated the `help...` methods in a00bb1f999214d81da6461af7779dfe7261b8ece.

File: WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginEmailFragment.java [hunk 4]
```diff
@@ -255,4 +365,92 @@ public void onAvailabilityChecked(OnAvailabilityChecked event) {
...
+        if (event.isError()) {
+            disconnectGoogleClient();
+            AppLog.e(T.API, "LoginEmailFragment.onAuthenticationChanged: " + event.error.type + " - " + event.error.message);
+            AnalyticsTracker.track(AnalyticsTracker.Stat.LOGIN_FAILED, event.getClass().getSimpleName(),
+                    event.error.type.toString(), event.error.message);
+            // TODO: Show error screen.
+        } else if (mAccountStore.hasAccessToken()) {
+            mDispatcher.dispatch(AccountActionBuilder.newFetchAccountAction());
```
[the remark this answers is not in the document]
Comment (theck13) [reply WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginEmailFragment.java]: That condition was mimicking [`SIgnInFragment`](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/accounts/SignInFragment.java#L1228-L1231).  I updated to `doFinishLogin` in 47fda3625e59706318916ad9edae1c1eaf95c326.

[the remark this answers is not in the document]
Comment (hypest) [reply WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginEmailFragment.java]: Ah, I see. Yeah, `SignInFragment` is on its way out so, it's not a good role model anymore.

**Comment by hypest** [comment]
Let's merge this and we can work on a couple more things on the next one: The progress dialog (on `onAuthenticationChanged`) is never dismissed and its message is probably not too clear.

LGTM!

**Pull request merged**